specs/latest/1.0/index.html

```diff
@@ -443,6 +448,46 @@ <h3><a name="THE_DRAWING_BUFFER">The Drawing Buffer</a></h3>
             generate <code>INVALID_VALUE</code>.
 
             If allocation fails, generate <code>OUT_OF_MEMORY</code>.
+        </dd>
+
+        <dt class="idl-code">undefined drawingBufferToneMapping(WebGLToneMapping toneMapping);
+        <dd>
+            <p>
+              The value of <code>toneMapping.mode</code> indicates how the drawing buffer shall be
+              tone mapped by the HTML page compositor when presented.
```

Comment:
For MSAA contexts, does the projection apply before or after the multisample resolve?

If I recall correctly, current implementations have the problem that already the resolve is done in incorrect colorspace (was it done in linear, even though the end result is used as srgb?). Not sure if that factors into the exact pixel equation tone mapping is about to modify...

Reply:
Short answer: Tone mapping is done after multisample resolve.

Long answer: The "bad behavior" sometimes seen in this issue is related to color space conversion (namely, color conversion happening after the resolve), not tone mapping (which is after-resolve). In the spec we indicate that tone mapping is done after conversion to the color space of the screen, which is always post-resolve.

Really long answer (about color space conversion, not tone mapping): The resolve is most correctly done in a linear space (a space that has a linear relationship to physical light quantities).

The resolve is done on raw pixel values (with one exception). This means:
* By default, if you're using `RGBA8` as the pixel format, and `"srgb"` or `"display-p3"` as your color space, then you will get "the wrong values", because the pixel values are not linearly related to physical light.
* If the framebuffer is `SRGB8_ALPHA8`, then (IIUC), the resolve will happen after a conversion to linear space, so, you get "the right values".
* If the framebuffer is `RGBA16F` and your color space is `"srgb"` or `"display-p3"`, then you will get "the wrong values".
* If you specify a linear color space (like the below-mentioned `"rec2100-linear"`), then you will get "the right values", because the pixels themselves correspond to linear light values. This will probably be the most common case for WebGL and WebGPU HDR applications.

I'd like for us to add `"rec2100-linear"` as a `PredefinedColorSpace` very soon, but when we do it, I'll want to make sure we get some extra details right (namely, when converted to an image, which will have to be fixed-point, it should be PQ encoded). I think there is now strong consensus on what's "right" here, but I want those issues to be very-well-ventilated in their own feature proposals. See some more comments below.